Verify line pragmas on generated code

ajaybhargavb opened the pull request:
https://github.com/aspnet/AspNetCore/issues/6373

- ~Only verifying design time for now using Source mappings~

ajaybhargavb commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentBindLoweringPass.cs:
```diff
@@ -249,7 +249,10 @@ private IntermediateNode[] RewriteUsage(IntermediateNode parent, TagHelperProper
-                valueNode.Children.Add(new CSharpExpressionAttributeValueIntermediateNode());
+                valueNode.Children.Add(new CSharpExpressionAttributeValueIntermediateNode()
+                {
+                    Source = node.Source
```
This was missing the SourceSpan.

ajaybhargavb commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentDesignTimeNodeWriter.cs:
```diff
@@ -244,13 +244,12 @@ public override void WriteCSharpExpressionAttributeValue(CodeRenderingContext co
 
-            var firstChild = node.Children[0];
-            if (firstChild.Source != null)
+            if (node.Source != null)
```
I can't remember why we look at firstChild here. Does any one remember why?

ajaybhargavb commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/CodeGeneration/CodeRenderingContext.cs:
```diff
@@ -40,5 +39,9 @@ public abstract class CodeRenderingContext
 
         public abstract void RenderChildren(IntermediateNode node, IntermediateNodeWriter writer);
+
+        public virtual void AddLinePragma(LinePragma linePragma)
```
Already spoke to @NTaylorMullen about this. This needs to be public because we share `CodeWriterExtensions.cs` which uses this method using compile include in `Mvc.Razor.Extensions`.

rynowak replied:
seems fine.

ajaybhargavb commented on src/Razor/test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/IntegrationTests/ComponentDesignTimeCodeGenerationTest/AsyncEventHandler_OnElement_ActionEventArgs_Lambda/TestComponent.codegen.cs:
```diff
@@ -26,7 +26,13 @@ public class TestComponent : Microsoft.AspNetCore.Components.ComponentBase
         {
-            __o = Microsoft.AspNetCore.Components.EventCallback.Factory.Create<Microsoft.AspNetCore.Components.UIMouseEventArgs>(this, "async (e) => await Task.Delay(10)");
+#nullable restore
+#line 2 "x:\dir\subdir\Test\TestComponent.cshtml"
```
Line 2 of `TestComponent.cshtml` is 
`<input onclick=""async (e) => await Task.Delay(10)"" />`.
I assume I am correct in generating a line pragma here?

rynowak replied:
This isn't super correct. See how the lambda isn't padded correctly?

rynowak replied:
Also, I want to know how this bug got put into preview 6. This is a serious-pants regression.

NTaylorMullen replied:
https://github.com/aspnet/AspNetCore/issues/8826#issuecomment-478102932

rynowak commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/CodeGeneration/LinePragma.cs:
```diff
@@ -0,0 +1,60 @@
+
+namespace Microsoft.AspNetCore.Razor.Language.CodeGeneration
+{
+    public sealed class LinePragma : IEquatable<LinePragma>
```
readonly struct?

rynowak commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/CodeGeneration/LinePragma.cs:
```diff
@@ -0,0 +1,60 @@
+            return hashCodeCombiner;
+        }
+
+        public override string ToString()
```
Is the `ToString` just for testing?

ajaybhargavb replied:
Yes. Technically the entire type is just for testing.

rynowak commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentBindLoweringPass.cs:
```diff
@@ -264,7 +267,10 @@ private IntermediateNode[] RewriteUsage(IntermediateNode parent, TagHelperProper
-                changeNode.Children.Add(new CSharpExpressionAttributeValueIntermediateNode());
+                changeNode.Children.Add(new CSharpExpressionAttributeValueIntermediateNode()
+                {
+                    Source = node.Source
```
Do we know what the effect is of making two pieces of code have the same line mapping? What happens if you try to set a breakpoint?

rynowak replied:
You can simulate these things btw by just making a regular C# file with pragmas if you want to test how the debugger behaves.

ajaybhargavb replied:
I tried simulating a bunch of things in a regular cs file. If I put a breakpoint on the line that maps to two locations in the generated code, it only breaks on the first location unless the contents of the two locations are the same. In our case, the contents of valueNode and changeNode are different. So, I guess we need to split the expression in to multiple lines if we want correct debugging.

rynowak replied:
I think it would be OK to just make one of the nodes debuggable.

rynowak commented on src/Razor/test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/BasicComponent_DesignTime.codegen.cs:
```diff
@@ -32,7 +32,7 @@ protected override void BuildRenderTree(Microsoft.AspNetCore.Components.RenderTr
 #nullable restore
 #line 2 "TestFiles/IntegrationTests/CodeGenerationIntegrationTest/BasicComponent.cshtml"
-       __o = this.ToString();
+      __o = this.ToString();
```
What's the root cause of these things scooting over?

ajaybhargavb replied:
I was wondering the same thing on Friday. Hence my commit message says `(O_o)` 😄 
Investigating now.

rynowak commented on src/Razor/test/Microsoft.AspNetCore.Razor.Test.Common/Language/IntegrationTests/IntegrationTestBase.cs:
```diff
@@ -519,6 +519,37 @@ protected void AssertSourceMappingsMatchBaseline(RazorCodeDocument codeDocument)
+        {
+            if (FileName == null)
+            {
+                var message = $"{nameof(AssertSourceMappingsMatchBaseline)} should only be called from an integration test ({nameof(FileName)} is null).";
```
nit: period after `integration test`

rynowak commented on src/Razor/test/Microsoft.AspNetCore.Razor.Test.Common/Language/IntegrationTests/RazorBaselineIntegrationTestBase.cs:
```diff
@@ -154,6 +154,33 @@ protected void AssertCSharpDocumentMatchesBaseline(RazorCodeDocument codeDocumen
+            AssertLinePragmas(codeDocument);
+        }
+
+        protected void AssertLinePragmas(RazorCodeDocument codeDocument)
```
Did we find any bugs in old Razor?

ajaybhargavb replied:
Not for the design time tests atleast. This could also mean we missed a source mapping somewhere because that is our source of truth. But that's unlikely.

ajaybhargavb replied:
I did come across some inconsistencies with how we generate code for `@model` directives when investigating https://github.com/aspnet/AspNetCore/issues/8790. I need to dig more to see if we missed something.

ajaybhargavb commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/CodeGeneration/DefaultCodeRenderingContext.cs:
```diff
@@ -94,6 +94,8 @@ internal class DefaultCodeRenderingContext : CodeRenderingContext
 
         public List<SourceMapping> SourceMappings { get; }
 
+        internal List<LinePragma> LinePragmas { get; }
```
This and `SourceMappings` above are not really exposed. Note the absence of `override`. This is just an internal implementation detail.

ajaybhargavb commented:
🆙 📅 No longer a draft PR
- Verify line pragmas in runtime tests as well
- Fixed a lot of places we didn't generate line pragmas
- Addressed feedback

ajaybhargavb commented on src/Razor/test/Microsoft.AspNetCore.Razor.Test.Common/Language/IntegrationTests/IntegrationTestBase.cs:
```diff
@@ -519,6 +519,70 @@ protected void AssertSourceMappingsMatchBaseline(RazorCodeDocument codeDocument)
+                {
+                    var content = sourceContent.Substring(classifiedSpan.Span.AbsoluteIndex, classifiedSpan.Span.Length);
+                    if (!string.IsNullOrWhiteSpace(content) &&
+                        classifiedSpan.BlockKind != BlockKindInternal.Directive &&
```
Excluding directives here because there are some inconsistencies here and there. E.g,
For `@section SectionName { ... }`, `SectionName` is categorized as code and there is no corresponding line pragma in generated code. Similarly for `@implements`. I'll work on fixing them incrementally after this PR.

rynowak replied:
It seems good for section. It seems like we want to fix things like `implements`.

ajaybhargavb commented on src/Razor/test/Microsoft.AspNetCore.Razor.Test.Common/Language/IntegrationTests/IntegrationTestBase.cs:
```diff
@@ -519,6 +519,70 @@ protected void AssertSourceMappingsMatchBaseline(RazorCodeDocument codeDocument)
+            Assert.NotNull(csharpDocument);
+            var linePragmas = csharpDocument.LinePragmas;
+            designTime = false;
+            if (designTime)
```
I am thinking of removing this entire block and using classified spans as the source of truth for design time tests as well. Thoughts?

rynowak replied:
Is the idea that we'd use the spans instead of the source mappings?

What I'd want to know is "do we validate the spans and source mappings together"?

ajaybhargavb replied:
> Is the idea that we'd use the spans instead of the source mappings?

Yes

> What I'd want to know is "do we validate the spans and source mappings together"?

Also yes. So my plan is the following,
```
foreach code span:
    verify if a line pragma exists
    if DesignTime:
        verify if a corresponding source mapping exists
```

ajaybhargavb replied:
Nevermind. This won't work because there isn't always a 1:1 mapping between SourceMappings and LinePragmas. I am going to stick to what's here already.

rynowak commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentDesignTimeNodeWriter.cs:
```diff
@@ -244,54 +244,20 @@ public override void WriteCSharpExpressionAttributeValue(CodeRenderingContext co
+            for (var i = 0; i < node.Children.Count; i++)
             {
-                using (context.CodeWriter.BuildLinePragma(firstChild.Source.Value, context))
+                if (node.Children[i] is IntermediateToken token && token.IsCSharp)
```
Don't we need the padding here? I see that it was removed

ajaybhargavb replied:
`WriteCSharpToken` takes care of adding the padding.

rynowak commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentEventHandlerLoweringPass.cs:
```diff
@@ -163,7 +163,10 @@ private IntermediateNode RewriteUsage(IntermediateNode parent, TagHelperProperty
                 }
 
-                result.Children.Add(new CSharpExpressionAttributeValueIntermediateNode());
+                result.Children.Add(new CSharpExpressionAttributeValueIntermediateNode()
```
This is worth a comment or something explaining what the strategy is.

rynowak commented on src/Razor/test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/BasicComponent_DesignTime.codegen.cs:
```diff
@@ -30,13 +30,15 @@ public class AspNetCore_d3c3d6059615673cb46fc4974164d61eabadb890 : Microsoft.Asp
 #nullable restore
 #line 2 "TestFiles/IntegrationTests/CodeGenerationIntegrationTest/BasicComponent.cshtml"
-       __o = this.ToString();
+             this.ToString()
```
❤️
